core/src/traits.rs

```diff
@@ -105,6 +106,10 @@ where
 	to_rpc_params_impl!();
 }
 
+impl ToRpcParams for Value {
```

Comment:
/cc @jsdw maybe it does more sense to implement this for P: Serialize to be generic?

We already implement this for vec, slices and arrays....

Reply:
I tried IIRC but there was an error, but yes I think this would be a really good improvment

Reply:
After looking at this again, it's technically possible that the serde_json::Value isn't representing as an array or object.

For instance `serde_json::Value::String("Foo")` will be serialized to be "Foo" which isn't valid JSON-RPC params it must be array or object.

So, can you please write a specialized `ToRpcParams` impl instead of using the serializing decl macro?


I think we could something similar to this:

```rust
impl ToRpcParams for serde_json::Value {
	fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
		match self {
			serde_json::Value::Array(a) => RawValue::from_string(serde_json::to_string(&a)?).map(Some),
			serde_json::Value::Object(o) => RawValue::from_string(serde_json::to_string(&o)?).map(Some),
			serde_json::Value::Null => Ok(None),
			serde_json::Value::Bool(b) => RawValue::from_string(serde_json::to_string(&[b])?).map(Some),
			serde_json::Value::String(s) => RawValue::from_string(serde_json::to_string(&[s])?).map(Some),
			serde_json::Value::Number(n) => RawValue::from_string(serde_json::to_string(&[n])?).map(Some),
		}
	}
}
```

Reply:
The only thing with `Ok(None)` iirc is that it's to indicate that no params have been provided, which might be confusing if some were. Really, I guess it should result in an error.

You _could_ implement `ToRpcParams` on eg https://docs.rs/serde_json/latest/serde_json/struct.Map.html though, which should always be valid?

Reply:
> The only thing with Ok(None) iirc is that it's to indicate that no params have been provided, which might be confusing if some were. Really, I guess it should result in an error.

We could also just serialize it and leave it to user to deal with it but yeah `Ok(None)` is bad.

> You could implement ToRpcParams on eg https://docs.rs/serde_json/latest/serde_json/struct.Map.html though, which should always be valid?

Yes

Reply:
Ok, the conclusion is that we are okay to implement this for [serde_json::Map](https://docs.rs/serde_json/latest/serde_json/struct.Map.html) but the implementation for `serde_json::Value` becomes to opinionated and better that users wrap in a new type pattern implement that trait themselves.